File: src/lib/canvas-geometry.ts
import type { TableInfo } from './ir';

export interface CanvasTableLike {
	alias: string;
	x: number;
	y: number;
	width?: number;
	height?: number;
	tableInfo: TableInfo;
}

export const CANVAS_CARD_WIDTH = 224;
export const CANVAS_CARD_HEIGHT = 297;
export const CANVAS_HEADER_HEIGHT = 41;
export const CANVAS_ROW_HEIGHT = 28;

/** Top of the row containing `columnName` inside the given table card. */
export function columnY(
	ct: CanvasTableLike,
	columnName: string,
	cardWidth = CANVAS_CARD_WIDTH,
	headerHeight = CANVAS_HEADER_HEIGHT,
	rowHeight = CANVAS_ROW_HEIGHT
): number {
	const idx = ct.tableInfo.columns.findIndex((c) => c.name === columnName);
	return ct.y + headerHeight + idx * rowHeight + rowHeight / 2;
}

/** Right edge of the card for the given table. */
export function tableRightX(ct: CanvasTableLike, cardWidth = CANVAS_CARD_WIDTH): number {
	return ct.x + (ct.width ?? cardWidth);
}

/** Left edge of the card for the given table. */
export function tableLeftX(ct: CanvasTableLike): number {
	return ct.x;
}

/** Cubic Bezier path between two tables' join endpoints. */
export function joinEdgePath(
	left: CanvasTableLike,
	leftColumn: string,
	right: CanvasTableLike,
	rightColumn: string,
	cardWidth = CANVAS_CARD_WIDTH
): string {
	const sourceX = tableRightX(left, cardWidth);
	const targetX = tableLeftX(right);
	const sourceY = columnY(left, leftColumn, cardWidth);
	const targetY = columnY(right, rightColumn, cardWidth);
	const midX = (sourceX + targetX) / 2;
	return `M ${sourceX} ${sourceY} C ${midX} ${sourceY}, ${midX} ${targetY}, ${targetX} ${targetY}`;
}

File: compose-app/shared/src/main/kotlin/com/safedb/query/CanvasGeometry.kt
package com.safedb.query

import com.safedb.model.TableInfo

data class CanvasTableLike(
    val alias: String,
    val x: Float,
    val y: Float,
    val width: Float? = null,
    val height: Float? = null,
    val tableInfo: TableInfo,
)

const val CANVAS_CARD_WIDTH = 224f
const val CANVAS_CARD_HEIGHT = 297f
const val CANVAS_HEADER_HEIGHT = 41f
const val CANVAS_ROW_HEIGHT = 28f

const val MIN_TABLE_WIDTH = 180f
const val MAX_TABLE_WIDTH = 520f
const val MIN_TABLE_HEIGHT = 140f
const val MAX_TABLE_HEIGHT = 640f

fun clampDimension(value: Float, min: Float, max: Float): Float {
    if (!value.isFinite()) return min
    return value.coerceIn(min, max).let { kotlin.math.round(it) }
}

/** Top of the row containing [columnName] inside the given table card. */
fun columnY(
    ct: CanvasTableLike,
    columnName: String,
    cardWidth: Float = CANVAS_CARD_WIDTH,
    headerHeight: Float = CANVAS_HEADER_HEIGHT,
    rowHeight: Float = CANVAS_ROW_HEIGHT,
): Float {
    val idx = ct.tableInfo.columns.indexOfFirst { it.name == columnName }
    return ct.y + headerHeight + idx * rowHeight + rowHeight / 2f
}

fun tableRightX(ct: CanvasTableLike, cardWidth: Float = CANVAS_CARD_WIDTH): Float =
    ct.x + (ct.width ?: cardWidth)

fun tableLeftX(ct: CanvasTableLike): Float = ct.x

/** Cubic Bezier path between two tables' join endpoints (SVG path data). */
fun joinEdgePath(
    left: CanvasTableLike,
    leftColumn: String,
    right: CanvasTableLike,
    rightColumn: String,
    cardWidth: Float = CANVAS_CARD_WIDTH,
): String {
    val points = joinEdgePoints(left, leftColumn, right, rightColumn, cardWidth)
    return "M ${points.sourceX} ${points.sourceY} C ${points.control1X} ${points.control1Y}, ${points.control2X} ${points.control2Y}, ${points.targetX} ${points.targetY}"
}

data class JoinEdgePoints(
    val sourceX: Float,
    val sourceY: Float,
    val control1X: Float,
    val control1Y: Float,
    val control2X: Float,
    val control2Y: Float,
    val targetX: Float,
    val targetY: Float,
)

fun joinEdgePoints(
    left: CanvasTableLike,
    leftColumn: String,
    right: CanvasTableLike,
    rightColumn: String,
    cardWidth: Float = CANVAS_CARD_WIDTH,
): JoinEdgePoints {
    val sourceX = tableRightX(left, cardWidth)
    val targetX = tableLeftX(right)
    val sourceY = columnY(left, leftColumn, cardWidth)
    val targetY = columnY(right, rightColumn, cardWidth)
    val midX = (sourceX + targetX) / 2f
    return JoinEdgePoints(
        sourceX = sourceX,
        sourceY = sourceY,
        control1X = midX,
        control1Y = sourceY,
        control2X = midX,
        control2Y = targetY,
        targetX = targetX,
        targetY = targetY,
    )
}
